rcl/src/rcl/event.c

```diff
@@ -0,0 +1,182 @@
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  // Allocate space for the implementation struct.
+  event->impl = (rcl_event_impl_t *) allocator->allocate(
+    sizeof(rcl_event_impl_t), allocator->state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    event->impl, "allocating memory failed", ret = RCL_RET_BAD_ALLOC; return ret);
+
+  event->impl->rmw_handle = rmw_get_zero_initialized_event();
+  event->impl->allocator = *allocator;
+
+  rmw_event_type_t rmw_event_type = RMW_EVENT_INVALID;
+  switch (event_type) {
+    case RCL_PUBLISHER_OFFERED_DEADLINE_MISSED:
+      rmw_event_type = RMW_EVENT_OFFERED_DEADLINE_MISSED;
+      break;
+    case RCL_PUBLISHER_LIVELINESS_LOST:
+      rmw_event_type = RMW_EVENT_LIVELINESS_LOST;
+      break;
+    default:
+      return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  return rmw_publisher_event_init(&event->impl->rmw_handle,
+    publisher->impl->rmw_handle, rmw_event_type);
+}
+
+rcl_ret_t
+rcl_subscription_event_init(
+  rcl_event_t * event,
+  const rcl_subscription_t * subscription,
+  const rcl_subscription_event_type_t event_type)
+{
+  rcl_ret_t ret = RCL_RET_OK;
+
+  // Check subscription and allocator first, so allocator can be used with errors.
+  RCL_CHECK_ARGUMENT_FOR_NULL(subscription, RCL_RET_INVALID_ARGUMENT);
+  rcl_allocator_t * allocator = &subscription->impl->options.allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  // Allocate space for the implementation struct.
+  event->impl = (rcl_event_impl_t *) allocator->allocate(
+    sizeof(rcl_event_impl_t), allocator->state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    event->impl, "allocating memory failed", ret = RCL_RET_BAD_ALLOC; return ret);
+
+  event->impl->rmw_handle = rmw_get_zero_initialized_event();
+  event->impl->allocator = *allocator;
+
+  rmw_event_type_t rmw_event_type = RMW_EVENT_INVALID;
+  switch (event_type) {
+    case RCL_SUBSCRIPTION_REQUESTED_DEADLINE_MISSED:
+      rmw_event_type = RMW_EVENT_REQUESTED_DEADLINE_MISSED;
+      break;
+    case RCL_SUBSCRIPTION_LIVELINESS_CHANGED:
+      rmw_event_type = RMW_EVENT_LIVELINESS_CHANGED;
+      break;
+    default:
+      return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  return rmw_subscription_event_init(&event->impl->rmw_handle,
+    subscription->impl->rmw_handle, rmw_event_type);
+}
+
+rcl_ret_t
+rcl_take_event(
+  const rcl_event_t * event,
+  void * event_info)
+{
+  bool taken = false;
+  RCL_CHECK_ARGUMENT_FOR_NULL(event, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(event_info, RCL_RET_INVALID_ARGUMENT);
+  rmw_ret_t ret = rmw_take_event(&event->impl->rmw_handle, event_info, &taken);
+  if (RMW_RET_OK != ret) {
+    RCL_SET_ERROR_MSG(rmw_get_error_string().str);
+    return rcl_convert_rmw_ret_to_rcl_ret(ret);
+  }
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Event take request succeeded: %s", taken ? "true" : "false");
```

Comment:
Please try to avoid memory allocation where possible (especially in commonly called functions like publish and take, as opposed to init and fini), instead of doing a formatted string here, just use a generic message, or and if/else to log two different messages.

Reply:
Refactored